A91-outlier-detection-metrics.md

```diff
@@ -0,0 +1,73 @@
+A91: gRPC Metrics for Outlier Detection
+---
+* Author(s): @pardhukonakanchi, @huntsman90
+* Approver: @markdroth
+* Status: In Review
+* Implemented in: Core, Go
+* Last updated: 2025-03-05
+* Discussion at: https://groups.google.com/g/grpc-io/c/iMezDbq5U-g
+
+## Abstract
+
+This document proposes some new metrics that will be added in gRPC for xDS Client Outlier Detection.
+
+## Background
+
+[A50: gRPC xDS Outlier Detection Support][A50] is a spec for gRPC to support Outlier Detection. The current implementation only offers debug and trace logging in terms of visibility, which can be insufficient for understanding and diagnosing decision making in large-scale production systems. 
+
+Using [A79: Non-per-call Metrics Architecture][A79], it is possible to add granular metrics to make visibility into outlier detection easy for service owners utilizing gRPC.
+
+### Related proposals: 
+* [A50: gRPC xDS Outlier Detection Support][A50]
+* [A66: OpenTelemetry Metrics][A66]
+* [A79: Non-per-call Metrics Architecture][A79]
+* [A75: xDS Aggregate Cluster Behavior Fixes][A75]
+* [A89: Backend Service Metric Label][A89] (pending)
+
+[A50]: A50-xds-outlier-detection.md
+[A66]: A66-otel-stats.md
+[A75]: A75-xds-aggregate-cluster-behavior-fixes.md
+[A79]: A79-non-per-call-metrics-architecture.md
+[A89]: https://github.com/grpc/proposal/pull/471
+
+## Proposal
+
+[A79]’s non-per-call metrics architecture fits perfectly into the metrics reporting solution. We propose to aim for parity where appropriate with [Envoy’s Outlier Detection metric collection](https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#outlier-detection-statistics) to be collected in gRPC.
+
+Outlier Detection metrics will have the following labels:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | Indicates the target we are running outlier detection on, as described in [A66]. |
+| grpc.lb.backend_service | optional | The backend service to which the traffic is being sent, as described in [A89]. Note that this label will be supported only if [A75] has already been implemented. |
```

Comment:
`backend_service` isn't available to Outlier Detection, even with A75 and A89. A89 only passes it to per-RPC metrics. In this gRFC, I think we'll have to make a resolver attribute to pass down from non-aggregate cluster `cds` policies, similar to [the attribute passed from weighted_target](https://github.com/grpc/proposal/blob/413b966783781e98b69ae0d8025cff84b5f5962a/A78-grpc-metrics-wrr-pf-xds.md?plain=1#L77).

@markdroth, @dfawley, thoughts?

Reply:
A89 also passes down an attribute to be used in the WRR metrics, which are also not per-RPC.  I think we can use that same attribute for the OD metrics.  However, I agree that in order to make that attribute visible to the OD policy, we need to set the attribute in the cds policy instead of the xds_cluster_impl policy -- but we can't do that until A75 is implemented.

For more details, see my reply in https://github.com/grpc/proposal/pull/471#discussion_r2025304376.

Reply:
Sounds good.

@PardhuKonakanchi, please add a note at the end of the "Proposal" section saying that the implementation will populate this label based on the resolver attribute passed down from the cds policy, as described in A89.  Thanks!